src/Raven.Server/SchemaValidation/SchemaValidatorCache.cs

```diff
@@ -0,0 +1,142 @@
+                    continue;
+            }
+
+            var schemaValidator = new SchemaValidator(validator.Disabled)
+            {
+                SchemaDefinition = validator.SchemaDefinition
+            };
+
+            try
+            {
+                var blittable = _context.Value.Sync.ReadForMemory(validator.SchemaDefinition, "schema-validation");
+                if (blittable.TryGet(SchemaValidatorConstants.AdditionalProperties, out object additionalProperties) && additionalProperties is false &&
+                    blittable.TryGet(SchemaValidatorConstants.Properties, out BlittableJsonReaderObject properties) &&
+                    properties.Contains(Client.Constants.Documents.Metadata.Key) == false)
+                {
+                    properties.Modifications = new DynamicJsonValue(properties) { [Client.Constants.Documents.Metadata.Key] = new DynamicJsonValue() };
+
+                    blittable.Modifications = new DynamicJsonValue(blittable) { [SchemaValidatorConstants.Properties] = properties };
+
+                    using (_ = blittable)
+                        blittable = _context.Value.ReadObject(blittable, "modified-schema-validation");
+                }
+
+                schemaValidator.Init(blittable);
+            }
+            catch (Exception e)
+            {
+                if (_logger.IsErrorEnabled)
+                    _logger.Error($"Failed to parse the schema validator for collection {collection}", e);
+
+                continue;
+            }
+
+            schemaValidatorsToAdd ??= new List<(string, SchemaValidator)>();
+            schemaValidatorsToAdd.Add((collection, schemaValidator));
+        }
+
+        Dictionary<string, SchemaValidator> newSchemaValidators = null;
+
+        foreach (var existing in _schemaValidatorsPerCollection)
+        {
+            if (configuration.ValidatorsByCollection.ContainsKey(existing.Key) == false)
+            {
+                newSchemaValidators ??= new Dictionary<string, SchemaValidator>(_schemaValidatorsPerCollection);
+                newSchemaValidators.Remove(existing.Key);
+            }
+        }
+
+        if (schemaValidatorsToAdd != null)
+        {
+            newSchemaValidators ??= new Dictionary<string, SchemaValidator>(_schemaValidatorsPerCollection);
+
+            foreach (var keyValue in schemaValidatorsToAdd)
+            {
+                newSchemaValidators[keyValue.Collection] = keyValue.SchemaValidator;
+            }
+
+        }
+
+        if (newSchemaValidators != null)
+            _schemaValidatorsPerCollection = newSchemaValidators;
+    }
+
+    public void Validate(string collection, BlittableJsonReaderObject document, NonPersistentDocumentFlags nonPersistentFlags, DocumentsOperationContext context)
+    {
+        // TODO: check if we need to add more flags here
+        if (nonPersistentFlags.Contain(NonPersistentDocumentFlags.FromReplication) ||
```

Comment:
e.g. backups 
if schema was changed later on, and old documents are already there, you will not block the restore, right? For me it sounds like a problem because we are restoring, not importing (for import we want that IMO)